### ISSUE TITLE
Pyright fails to infer types assigned with SQLAlchemy's Mapped anotation

### ISSUE THREAD

**Issue by igsp7** [microsoft/pyright#9741]
**Describe the bug**
Pyright does not infer attributes and their types for classes representing SQLAlchemy models. When creating an instance of a SQLAlchemy model:
- Mandatory fields (e.g., nullable=False) are not validated, and Pyright does not complain if they are missing.
- Field types are not enforced. For example, passing an int to a field annotated as Mapped[str] does not raise any type error.

If I explicitly provide an `__init__()` function for the model, Pyright is able to provide typing hints and enforce types. However, this feels redundant and I am hopping that Pyright is able to infer types and enforce validation based on the Mapped type annotations and the SQLAlchemy model definition without requiring a manually written `__init__()`.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
from sqlalchemy.orm import Mapped, mapped_column, DeclarativeBase

class Base(DeclarativeBase):
    pass

class Seller(Base):
    __tablename__ = "sellers"
    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    company_name: Mapped[str]

seller = Seller() # Here I would expect pyright to complain about missing attributes but this is not happening. 
```

I am using `SQLAlchemy==2.0.37`

**VS Code extension or command-line**
pyright from command line: Version: 1.1.392.post0

**Comment by erictraut** [comment]
Pyright is a [standards-compliant](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) type checker, and it has no knowledge of SQLAlchemy or any other third-party library. It bases its analysis and error reporting on the type information provided to it (e.g. through annotations in source and stub files) and on typing rules spelled out in the Python type system.

I don't see any type rules that are violated in the code above. Is there something specific that you have in mind here?

I'm not a SQLAlchemy user, so I'm not familiar with what "missing attributes" you're referring to above. If you want pyright (or any standards-compliant Python type checker) to validate that values assigned to the variable `seller` conform to a particular type (either nominal or structural), you would need to provide a type declaration for `seller`. For example, if you want the type checker to validate that the assigned value has an attribute named `nullable` that is of a particular type, you would need to define a protocol that includes this attribute and then use this protocol in a type declaration.

```python
class Nullable(Protocol):
    nullable: bool

seller: Nullable = Seller() # Type violation
```

For more details about type declarations, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts).

For more details about the Python type system and the rules that type checkers use for determining type compatibility, refer to the [Python type specification](https://typing.readthedocs.io/en/latest/spec/index.html).

**Comment by igsp7** [comment]
Thanks a lot for the reply, @erictraut. I was expecting Pyright to complain about the missing `company_name` attribute when instantiating the object. Perhaps the issue lies in how the SQLAlchemy library dynamically generates the `__init__()` method, so I will open an issue with SQLAlchemy to explore this further.
